### DISCUSSION TITLE
Difference between partial rendering(PPR) and the current streaming suspense?

### DISCUSSION THREAD

**Discussion by jovanhartono** [vercel/next.js#58322] [Help] [answered]
next.js 14 introduces new partial prerendering(PPR) feature.

what's the difference with the current streaming and suspense?

https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming

**Comment by icyJoseph** [comment]
The Partial Rendering, is, what I think Suspense, should've been all along... 

It applies at build time, and what it does is, pre-render everything that's not bounded by a Suspense tag. 

That means that these parts are effectively part of the HTML, just like `getStaticProps`, or Automatic Static Optimization would do, but! when the page is served, the server will send the data that goes in between the Suspense boundaries. So you gain:

- Extremely fast Time to Visible, comparable to a page where everything is know at build time
- Still have the option to have dynamic areas within the page, which will receive their data from the React Server

So for example, as an extreme simplification, you could pre-render various parts of a user dashboard, and then the React Server makes sure to place the user data where the Suspense boundaries were, whereas, before you'd need to defer rendering anything in the user dashboard, to on-demand rendering, ala `getServerSideProps`.

Another example, a blog post, with a views counter. You'd need to either, render the blog post on demand, or pre-render it, and carve out an area where you'd use `useState + useEffect` to fetch the views count. With Partial Pre-rendering, you can pre-render the blog post, leave the view count between a Suspense boundary, and then at runtime, the React Server injects the views count where the Suspense boundary was. And you can even fetch the views count purely server side!

**Comment by Connor-Skudlarek** [comment, marked as answer]
PPR lets you render the entire page at build time, with the exception of Suspense wrapped components (likely other exceptions, but specific to this question that's the difference).

With streaming/suspense, you render the entire page _on request_. Suspense lets you fetch data in parallel, and streaming lets you replace the suspense component with the fully rendered bits as each of them finish. But everything is done on request time.

With PPR, you render part of the page at _build time_ and part of it at request. Anything that can be rendered at build time, is rendered at build time. Everything else should be wrapped in Suspense (with a fallback to prevent jarring CSS behavior).

So instead of needing to render the entire page (the current setup) or generate the entire page at build time (static pages), you can generate a "partial first page" at build time and render out the parts that are dynamic _while_ the server sends that static page. Everything about the app that doesn't change gets to the user _fast_.

This effectively gets the first paint more quickly, which feels faster even if the actual/real improvements in speed are negligible in some particular app. The user who sees things happening in 100ms increments is usually happier than the one waiting 700ms for something to happen, even if they both waited a total of 700ms.

**Reply by lbittner-pdftron** [reply]
Hey! Stumbled upon this conversation recently and have a question.

I've been experimenting with PPR and am finding that even the parts of my app that are supposed to be "static" are still writing logs to the server console on request. 

This can be reproduced easily by building a simple page like this

```
function Page() {
  console.log('Render')
 
 return (
   <div>
       <h1>Hello world</h1>
       <Suspense>
            <SomeComponentThatFetchesData />
       </Suspense>
   </div>
 )
}
```

When I build the project and request this page, I am still seeing the "Render" console log on the server.

Is this expected? I personally wouldn't expect something "static" to ever be executed server side.

**Reply by ssijak** [reply]
Can you emulate PPR with current Streaming/suspense by setting cache headers on static parts?

**Reply by paulgrieselhuber** [reply]
This is the best explanation of these concepts that I've seen. We linked to your explanation from our video on the subject: https://youtu.be/hdaE94dP_Vg

**Reply by Connor-Skudlarek** [reply]
@paulgrieselhuber Yooo! That's so cool. I'm glad it has helped those who have found this. 😄

I do want to amend some of what I said as there seems to be confusion over server rendering and client rendering, and what that means for React. Sorry @lbittner-pdftron if I've misled you. 

Worth keeping in mind that the server is rendering ahead of time to reduce the time it takes to send HTML to the client. It isn't preventing the need to render at all.

Even with SSG, where all of the HTML was rendered and nothing on the page is dynamic content. On request, the client will render at least once for React to complete hydration of the page (add event listeners for instance).

If the client rendered page is the same as the server rendered SSG (I'm pretty sure!) the virtual DOM won't update as there is no difference. But because it ran through the function, it will still console.log() during the render.

**Reply by ahmad-shoja** [reply]
That is a good answer, so clear thanks

**Reply by UltimateGG** [reply]
@lbittner-pdftron Did you ever find out why this is? I just did the exact same thing testing PPR and its very frustrating I feel like I don't get it or that its doing nothing. My best guess is its running this code still but only using the rendered output of the components in suspense?

**Comment by aarondentrodev** [comment]
While @Connor-Skudlarek 's answer is correct, I'm just going to describe what I understood about the PPR.

Without PPR, there are only 2 options to render the entire page:
- Static rendering (render the entire page at build time)
- Dynamic rendering (render the entire page for each user at request time)

As mentioned [here](https://nextjs.org/docs/app/building-your-application/rendering/server-components#switching-to-dynamic-rendering) and [here](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic),  Next.js uses static rendering by default and switch to dynamic rendering if any dynamic function or uncached data request is discovered during rendering or if the route or page is configured to force dynamic rendering.

Streaming is a technique to progressively render UI and only works for dynamic rendering. **PPR is a new concept that gives us a 3rd option to render the entire page by combining static rendering and dynamic rendering**, which means part of the page is rendered at build time for a fast initial load, while the remaining part is rendered dynamically at request time with streaming.

So I think they are different concepts but look similar because both of them use React's Suspense.

**Comment by magnusriga** [comment]
Just to add to the confusion here: I used a third party cache handler and checked out the cache, and even with PPR turned off and every layout and page behind a dynamic segment (`[folder]`), the pages are still placed in the cache on each visit.

Moreover, also with PPR turned off, these cached pages have placeholders for dynamic components (those wrapped in suspense).

As such, I have to admit I do not currently see a lot of benefit of PPR. Sure, it will be slightly faster for the very first user that visits a page, but for everyone else it seems unnecessary. I might of course be missing something, but this is what I see when inspecting the cache.
